File: Hard/124. Binary Tree Maximum Path Sum/maxPathSum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
/*

    Time complexity O(n)
    Space complexity O(h)
*/
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        if(!root){
            return 0;
        }
        int ans = INT_MIN;
        maxPathSum(root, ans);
        return ans;
    }
private:
    int maxPathSum(TreeNode* root, int &ans){
        if(!root){
            return 0;
        }
        int l = max(maxPathSum(root->left, ans), 0); //負值就不取了
        int r = max(maxPathSum(root->right, ans), 0);
        int sum = root->val + l + r;
        ans = max(ans, sum); //更新ans
        return max(l, r) + root->val;  //返回的話要返回單一路徑 我們取最大的
    }
};

File: Hard/124. Binary Tree Maximum Path Sum/maxPathSum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        if(root == null){
            return 0;
        }
        maxPath(root);
        return ans;
    }
    private int maxPath(TreeNode root){
        if(root == null){
            return 0;
        }
        int l = Math.max(0, maxPath(root.left));
        int r = Math.max(0, maxPath(root.right));
        int sum = l + r + root.val;
        ans = Math.max(ans, sum);
        return Math.max(l, r) + root.val;
    }
}